src/modeling_library/product.jl

```diff
@@ -0,0 +1,104 @@
+These arguments are the arguments to each component distribution, in the order in which the distributions are passed to the constructor.
+
+Example:
+```julia
+@gen function unit_strip_and_near_seven()
+    x ~ flip_and_number(0.0, 0.1, 7.0, 0.01)
+end
+```
+"""
+struct ProductDistribution{T} <: Distribution{T}
+    K::Int
+    distributions::Vector{<:Distribution}
+    has_output_grad::Bool
+    has_argument_grads::Tuple
+    is_discrete::Bool
+    num_args::Vector{Int}
+    starting_args::Vector{Int}
+end
+
+(dist::ProductDistribution)(args...) = random(dist, args...)
+
+Gen.has_output_grad(dist::ProductDistribution) = dist.has_output_grad
+Gen.has_argument_grads(dist::ProductDistribution) = dist.has_argument_grads
+Gen.is_discrete(dist::ProductDistribution) = dist.is_discrete
+
+function ProductDistribution(distributions::Vararg{<:Distribution})
+    _has_output_grads = true
+    _is_discrete = true
+
+    types = Type[]
+
+    _has_argument_grads = Bool[]
+    _num_args = Int[]
+    _starting_args = Int[]
+    start_pos = 1
+
+    for dist in distributions
+        type = typeof(dist)
+        while supertype(type) != Any
+            type = supertype(type)
+        end
+        push!(types, type.parameters[1])
+
+        _has_output_grads = _has_output_grads && has_output_grad(dist)
+        _is_discrete = _is_discrete && is_discrete(dist)
+
+        grads_data = has_argument_grads(dist)
+        append!(_has_argument_grads, grads_data)
+        push!(_num_args, length(grads_data))
+        push!(_starting_args, start_pos)
+        start_pos += length(grads_data)
+    end
+
+    return ProductDistribution{Tuple{types...}}(
+        length(distributions),
+        collect(distributions),
+        _has_output_grads,
+        Tuple(_has_argument_grads),
+        _is_discrete,
+        _num_args,
+        _starting_args)
+end
+
+function extract_args_for_component(dist::ProductDistribution, component_args_flat, k::Int)
+    start_arg = dist.starting_args[k]
+    n = dist.num_args[k]
+    return component_args_flat[start_arg:start_arg+n-1]
+end
+
+Gen.random(dist::ProductDistribution, component_args_flat...) =
+    [random(dist.distributions[k], extract_args_for_component(dist, component_args_flat, k)...) for k in 1:dist.K]
+
+Gen.logpdf(dist::ProductDistribution, x, component_args_flat...) =
+    sum(Gen.logpdf(dist.distributions[k], x[k], extract_args_for_component(dist, component_args_flat, k)...) for k in 1:dist.K)
+
+function Gen.logpdf_grad(dist::ProductDistribution, x, component_args_flat...)
+    logpdf_grads = [Gen.logpdf_grad(dist.distributions[k], x[k], extract_args_for_component(dist, component_args_flat, k)...) for k in 1:dist.K]
+    x_grad = if dist.has_output_grad
+        tuple((grads[1] for grads in logpdf_grads)...)
+    else
+        nothing
+    end
+    arg_grads = vcat((collect(grads[2:end]) for grads in logpdf_grads)...)
+    return (x_grad, arg_grads...)
+end
```

Comment:
Similar to my comments above, I think there's a way to reduce both the chance of runtime dispatch and memory allocation in this function. In particular, I think this will allocate less if we don't keep creating new arrays to hold the gradients (which happens with list comprehensions and `collect`), and instead use a for loop. Here's my suggested implementation:

```julia
function Gen.logpdf_grad(dist::ProductDistribution, x, args...)
    x_grad = ()
    arg_grads = ()
    for (k, d) in enumerate(dist.distributions)
          (x_g, arg_gs...) = Gen.logpdf_grad(d, x[k], extract_args_for_component(dist, args, k)...)
          x_grad = (x_grad..., x_g)
          arg_grads = (arg_grads..., arg_gs)
    end
    x_grad = dist.has_output_grad ? x_grad : nothing
    return (x_grad, arg_grads...)
end
```

I _think_ the Julia compiler will optimize this to something more efficiently (and mostly non-allocating), but I'm not a 100% sure, and it's probably worth testing.

Also, if it's too much work, I'm fine with leaving the implementation as is.

Reply:
I just pushed a tiny variant of your suggestion.  One thing I am curious about is whether `x_grad = (x_grad..., x_g)` constitutes an additional allocation or not.  In a language like C we could allocate the whole tuple upfront and fill in its values as we went...

Reply:
I'm not entirely sure without testing! If you use the `@time` macro, it'll usually tell you something about whether the function allocates (though it won't tell you which is responsible). Julia has a lot of profiling tools that help with this too, such as https://github.com/JuliaLang/AllocCheck.jl

Anyway, my understanding is that as long as Julia can infer the types of the variables like `x_grad`, statements like `x_grad = (x_grad..., x_g)` should be compiled into non-allocating code, as long as the original value of `x_grad` is also a tuple with a relatively small number of entries (there's some constant beyond which the Julia compiler will avoid the optimization).